File: candidature-hub/app/lib/antivirus.ts
import net from "node:net";

const host = process.env.CLAMAV_HOST || "clamav";
const port = Number(process.env.CLAMAV_PORT) || 3310;
const required = process.env.ANTIVIRUS_REQUIRED !== "0";

export type AntivirusResult = { clean: true } | { clean: false; threat: string };

export async function scanBuffer(buffer: Buffer): Promise<AntivirusResult> {
  if (!required) return { clean: true };
  return new Promise((resolve, reject) => {
    const socket = net.createConnection({ host, port });
    const chunks: Buffer[] = [];
    const fail = (error: Error) => reject(new Error(`Antivirus non disponibile: ${error.message}`));
    socket.setTimeout(30_000, () => socket.destroy(new Error("timeout")));
    socket.once("error", fail);
    socket.on("data", chunk => chunks.push(Buffer.from(chunk)));
    socket.once("connect", () => {
      socket.write("zINSTREAM\0");
      const size = Buffer.alloc(4);
      size.writeUInt32BE(buffer.length);
      socket.write(size);
      socket.write(buffer);
      socket.write(Buffer.alloc(4));
    });
    socket.once("end", () => {
      const response = Buffer.concat(chunks).toString("utf8").replace(/\0/g, "").trim();
      if (response.endsWith("OK")) return resolve({ clean: true });
      const match = response.match(/stream:\s+(.+)\s+FOUND/i);
      if (match) return resolve({ clean: false, threat: match[1] });
      reject(new Error(`Risposta antivirus non valida: ${response || "vuota"}`));
    });
  });
}

File: workers/antivirus.py
import os
import socket
import struct


class AntivirusUnavailable(RuntimeError):
    pass


class ThreatDetected(RuntimeError):
    pass


def scan_bytes(data: bytes, source: str = "file") -> None:
    if os.environ.get("ANTIVIRUS_REQUIRED", "1") == "0":
        return
    host = os.environ.get("CLAMAV_HOST", "clamav")
    port = int(os.environ.get("CLAMAV_PORT", "3310"))
    try:
        with socket.create_connection((host, port), timeout=15) as sock:
            sock.settimeout(30)
            sock.sendall(b"zINSTREAM\0")
            for offset in range(0, len(data), 1024 * 1024):
                chunk = data[offset:offset + 1024 * 1024]
                sock.sendall(struct.pack("!I", len(chunk)) + chunk)
            sock.sendall(struct.pack("!I", 0))
            response = b""
            while not response.endswith(b"\0"):
                part = sock.recv(4096)
                if not part:
                    break
                response += part
    except (OSError, socket.timeout) as error:
        raise AntivirusUnavailable(f"Antivirus non disponibile per {source}: {error}") from error
    result = response.rstrip(b"\0").decode("utf-8", "replace")
    if result.endswith(" OK"):
        return
    if result.endswith(" FOUND"):
        raise ThreatDetected(result)
    raise AntivirusUnavailable(f"Risposta antivirus non valida per {source}: {result or 'vuota'}")


def scan_file(path: str) -> None:
    with open(path, "rb") as handle:
        scan_bytes(handle.read(), os.path.basename(path))
